### PR TITLE
Public node: perf and fixes

### PR DESCRIPTION
**Pull request by maciejhirsz**
- Introduced a worker pool for webworkers, which speeds up re-generating of preview avatars (apparently each worker has to JIT the js on it's own, over and over again).
- All keythereum related stuff is now done in workers as well, so the UI never hangs.
- Regenerating the avatars on the new account modal now shows a spinner.
- Restoring accounts from private keys, testing passwords and changing passwords now works as intended.

### CODE REVIEW

File: js/src/api/local/accounts/account.js [hunk 1]
```diff
@@ -68,21 +70,23 @@ export default class Account {
...
+    return createKeyObject(key, password).then((keyObject) => {
+      this._keyObject = keyObject;
 
-    const keyObject = keythereum.dump(password, key, salt, iv);
+      this._persist();
+    });
+  }
 
-    const account = new Account(persist, { keyObject });
+  static fromPrivateKey (persist, key, password) {
```
Comment (jacogr) [inline js/src/api/local/accounts/account.js]: Not sure if you tested this 100%, if not there is a private key we use in tests on the Rust side inside tests (with a known account). If already done, ignore :)

File: js/src/api/local/accounts/accounts.js [hunk 2]
```diff
@@ -38,14 +38,22 @@ export default class Accounts {
 
   create (secret, password) {
     const privateKey = Buffer.from(secret.slice(2), 'hex');
-    const account = Account.fromPrivateKey(this.persist, privateKey, password);
 
-    this._store.push(account);
-    this.lastAddress = account.address;
+    return Account.fromPrivateKey(this.persist, privateKey, password)
```
Comment (jacogr) [inline js/src/api/local/accounts/accounts.js]: Small nitpick, would prefer 

```
return Account
  .fromPrivateKey(...)
  .then(...)
```

Just aligns with what we have in the rest of the codebase.

File: js/src/api/local/accounts/accounts.js [hunk 3]
```diff
@@ -73,28 +81,40 @@ export default class Accounts {
...
-    const index = this._store.findIndex((account) => account.address === address);
+    const account = this.get(address);
 
-    if (index === -1) {
+    if (!account) {
       return false;
     }
 
-    const account = this._store[index];
+    return account.isValidPassword(password)
```
Comment (jacogr) [inline js/src/api/local/accounts/accounts.js]: Same samll alignment comment as above.

File: js/src/api/local/ethkey/index.js [hunk 4]
```diff
@@ -14,31 +14,34 @@
...
-// Allow a web worker in the browser, with a fallback for Node.js
-const hasWebWorkers = typeof Worker !== 'undefined';
-const KeyWorker = hasWebWorkers ? require('worker-loader!./worker')
-                                : require('./worker').KeyWorker;
+import workerPool from './workerPool';
 
-// Local accounts should never be used outside of the browser
-export let keythereum = null;
+export function createKeyObject (key, password) {
+  return workerPool.getWorker().action('createKeyObject', { key, password })
```
Comment (jacogr) [inline js/src/api/local/ethkey/index.js]: As per above, prefer

```
return ...getWorker()
  .action(...)
  .then(...)
```

(Next function has it that way, just align)

File: js/src/api/local/ethkey/worker.js [hunk 5]
```diff
@@ -14,68 +14,117 @@
...
+
+      // Convert to array to safely send from the worker
+      return Array.from(key);
+    } catch (e) {
+      return null;
     }
   }
+};
+
+function bytesToHex (bytes) {
```
Comment (jacogr) [inline js/src/api/local/ethkey/worker.js]: Not 100% sure why this is "duplicated", i.e. not used from `/api/util/format`? (I looked at this the previous round as well, could see a slight difference, but non-duplication is obviously preferred)

File: js/src/api/local/middleware.js [hunk 6]
```diff
@@ -57,6 +57,21 @@ export default class LocalAccountsMiddleware extends Middleware {
       });
     });
 
+    register('parity_changePassword', ([address, oldPassword, newPassword]) => {
+      const account = accounts.get(address);
+
+      return account.decryptPrivateKey(oldPassword)
```
Comment (jacogr) [inline js/src/api/local/middleware.js]: Small nitpick as above.

**Comment by jacogr** [comment]
Generally looks good, one testing comment (probably already taken care of), some small style grumbles, just one function duplication where we can re-use. No major grumbles.

Generally though, think we should try and invest in tests for the middleware at some time sooner rather than later, just ensuring that what we do is correct. (And that the future holds to that promise)

File: js/src/modals/CreateAccount/NewAccount/newAccount.spec.js [hunk 7]
```diff
@@ -58,12 +58,13 @@ describe('modals/CreateAccount/NewAccount', () => {
         return instance.componentWillMount();
       });
 
-      it('creates initial accounts', () => {
-        expect(Object.keys(instance.state.accounts).length).to.equal(7);
+      it('resets the accounts', () => {
+        expect(instance.state.accounts).to.be.null;
+        // expect(Object.keys(instance.state.accounts).length).to.equal(7);
```
Comment (tomusdrw) [inline js/src/modals/CreateAccount/NewAccount/newAccount.spec.js]: Is it still needed?

File: js/src/api/local/accounts/account.js (hunk 1 above)
Comment (maciejhirsz) [inline js/src/api/local/accounts/account.js]: All the actual logic for this is in Keythereum, and the main issue here is that the tests take a while and can timeout in CI :|.

**Pull request merged**